### PR TITLE
gccrs: Implement C-style string literals

### PR DESCRIPTION
**Pull request by Polygonalr**
This is not part of Rust 1.49, but it's nice to backport it for now, since Rust for Linux uses C-style string literals a bit.

Current behaviour is similar to normal string literals, except they are type-checked as slice types instead of array, and the compiled fat pointer has +1 to size to include the null terminator.

### CODE REVIEW

File: gcc/rust/backend/rust-compile-expr.cc [hunk 1]
```diff
@@ -1900,6 +1904,31 @@ CompileExpr::compile_byte_string_literal (const HIR::LiteralExpr &expr,
...
+  bool ok = ctx->get_tyctx ()->lookup_builtin ("usize", &usize);
+  rust_assert (ok);
+  tree type = TyTyResolveCompile::compile (ctx, usize);
+
+  // +1 for null terminator, unlike Rust string literals.
+  tree size = build_int_cstu (type, literal_value.as_string ().size () + 1);
+
+  return Backend::constructor_expression (fat_pointer, false, {data, size}, -1,
+					  expr.get_locus ());
+}
```
Comment (CohenArthur) [inline gcc/rust/backend/rust-compile-expr.cc:1947]: I think we have to be careful here since C string literals are expected to produce references of `CStr` if I'm understanding things right

From https://doc.rust-lang.org/nightly/edition-guide/rust-2021/c-string-literals.html :

 > Literals of the form c"foo" or cr"foo" represent a string of type [&core::ffi::CStr](https://doc.rust-lang.org/nightly/core/ffi/struct.CStr.html).

So while the layout of a `CStr` and a `[c_char]` is the same...

```rust
#[derive(Hash)]
#[stable(feature = "rust1", since = "1.0.0")]
// FIXME:
// `fn from` in `impl From<&CStr> for Box<CStr>` current implementation relies
// on `CStr` being layout-compatible with `[u8]`.
// When attribute privacy is implemented, `CStr` should be annotated as `#[repr(transparent)]`.
// Anyway, `CStr` representation and layout are considered implementation detail, are
// not documented and must not be relied upon.
pub struct CStr {
    // FIXME: this should not be represented with a DST slice but rather with
    //        just a raw `c_char` along with some form of marker to make
    //        this an unsized type. Essentially `sizeof(&CStr)` should be the
    //        same as `sizeof(&c_char)` but `CStr` should be an unsized type.
    inner: [c_char],
}
```

I think we should still figure out how to convert to the expected type, otherwise we're going to have failures when compiling the kernel

File: gcc/rust/lex/rust-lex.cc [hunk 2]
```diff
@@ -1063,6 +1063,10 @@ Lexer::build_token ()
 	    return parse_raw_byte_string (loc);
 	}
 
+      // C-style strings
+      else if (current_char == 'c' && peek_input () == '"')
+	return parse_c_string (loc);
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:1068]: Note that in a later PR you can also add C-style raw strings the same way, but I don't know if the kernel makes use of those

File: gcc/rust/lex/rust-lex.cc [hunk 3]
```diff
@@ -1749,6 +1753,82 @@ Lexer::parse_byte_string (location_t loc)
...
+	}
+
+      current_column++;
+      if (current_char.value == '\n')
+	{
+	  current_line++;
+	  current_column = 1;
+	  // tell line_table that new line starts
+	  start_line (current_line, max_column_hint);
+	}
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:1803]: If this is the same code as other string parsing logic let's think about factoring it into a function

File: gcc/rust/parse/rust-parse-impl-expr.hxx [hunk 4]
```diff
@@ -2101,6 +2123,20 @@ Parser<ManagedTokenSource>::null_denotation_not_path (
       return std::unique_ptr<AST::LiteralExpr> (
 	new AST::LiteralExpr (tok->get_str (), AST::Literal::RAW_STRING,
 			      tok->get_type_hint (), {}, tok->get_locus ()));
+    case C_STRING_LITERAL:
+      if (flag_c_style_string_literals)
+	return std::unique_ptr<AST::LiteralExpr> (
+	  new AST::LiteralExpr (tok->get_str (), AST::Literal::C_STRING,
+				tok->get_type_hint (), {}, tok->get_locus ()));
```
Comment (CohenArthur) [inline gcc/rust/parse/rust-parse-impl-expr.hxx]: Formatting nitpick since we have an else block with brackets
```suggestion
      if (flag_c_style_string_literals)
    {
	return std::unique_ptr<AST::LiteralExpr> (
	  new AST::LiteralExpr (tok->get_str (), AST::Literal::C_STRING,
				tok->get_type_hint (), {}, tok->get_locus ()));
    }
```

**Review by CohenArthur — COMMENTED** [review]
I think the code is really good, but I'd like to have a think about that resulting `&CStr` type vs `[u8]` type before we merge it. Can you please check that as soon as C-style string literals were added, `CStr` became a lang item, in which case we should be able to refer to it within the typechecker and use that type, and if so, we have to figure out a way to get to the `CStr` type with a 1.49 libcore which does not have the `CStr` as a lang item yet

**Comment by dkm** [comment]
FWIW, the changelog is not correct. You need to follow the ` * file.ext (item): What changed` pattern.
Please do not simply point at the file :) See for example:

```
gcc/rust/ChangeLog:
	* lex/rust-token.h: Add C_STRING_LITERAL and RAW_C_STRING_LITERAL (unused
	for now).
	* lex/rust-lex.h: Define new parse_c_string function.
	* lex/rust-lex.cc: Implement lexing for C-style string literals.
	* hir/tree/rust-hir-literal.h: Define new C_STRING LitType.
```

Can you do some cleanup there? Thanks!

File: gcc/rust/backend/rust-compile-expr.cc (hunk 1 above)
Comment (Polygonalr) [reply gcc/rust/backend/rust-compile-expr.cc:1947]: Added `CStr` language item in my latest commit.

**Comment by Polygonalr** [comment]
> FWIW, the changelog is not correct. You need to follow the ` * file.ext (item): What changed` pattern. Please do not simply point at the file :) See for example:
> 
> ```
> gcc/rust/ChangeLog:
> 	* lex/rust-token.h: Add C_STRING_LITERAL and RAW_C_STRING_LITERAL (unused
> 	for now).
> 	* lex/rust-lex.h: Define new parse_c_string function.
> 	* lex/rust-lex.cc: Implement lexing for C-style string literals.
> 	* hir/tree/rust-hir-literal.h: Define new C_STRING LitType.
> ```
> 
> Can you do some cleanup there? Thanks!

@dkm Just did some commit message clean-up, apologies for not being specific enough in my commit messages :sweat:, I'll do my best not to repeat this in the future.

File: gcc/rust/util/rust-lang-item.cc [hunk 5]
```diff
@@ -92,6 +92,7 @@ const BiMap<std::string, LangItem::Kind> Rust::LangItem::lang_items = {{
   {"slice_u8", Kind::SLICE_U8},
   {"slice", Kind::SLICE},
   {"str", Kind::STR},
+  {"CStr", Kind::CSTR},
```
Comment (CohenArthur) [inline gcc/rust/util/rust-lang-item.cc:97]: This needs an extra comment (and possibly to be put at the end of the `BiMap`) with an explanation saying it's only used when compiling the Linux kernel with a patched core, and isn't present in Rust 1.49

File: gcc/rust/util/rust-lang-item.h [hunk 6]
```diff
@@ -125,6 +125,7 @@ class LangItem
     SLICE_U8,
     SLICE,
     STR,
+    CSTR,
```
Comment (CohenArthur) [inline gcc/rust/util/rust-lang-item.h:130]: Likewise

File: gcc/rust/backend/rust-compile-expr.cc [hunk 7]
```diff
@@ -1902,6 +1904,31 @@ CompileExpr::compile_byte_string_literal (const HIR::LiteralExpr &expr,
...
+  auto base = Backend::string_constant_expression (literal_value.as_string ());
+  tree data = address_expression (base, expr.get_locus ());
+
+  TyTy::BaseType *usize = nullptr;
+  bool ok = ctx->get_tyctx ()->lookup_builtin ("usize", &usize);
+  rust_assert (ok);
+  tree type = TyTyResolveCompile::compile (ctx, usize);
+
+  // +1 for null terminator, unlike Rust string literals.
+  tree size = build_int_cstu (type, literal_value.as_string ().size () + 1);
```
Comment (CohenArthur) [inline gcc/rust/backend/rust-compile-expr.cc:1943]: This is correct, but don't we also need to add an extra zero byte at the end?

Comment (CohenArthur) [reply gcc/rust/backend/rust-compile-expr.cc:1943]: Is it taken care of by the backend already?

**Review by CohenArthur — COMMENTED** [review]
Good work @Polygonalr :) well done

File: gcc/rust/backend/rust-compile-expr.cc (hunk 7 above)
Comment (Polygonalr) [reply gcc/rust/backend/rust-compile-expr.cc:1943]: [C++11 strings already come with a null terminator](https://stackoverflow.com/questions/6077189/is-stdstring-always-null-terminated-in-c11), so I don't think we need to handle it.

File: gcc/rust/lex/rust-lex.cc (hunk 3 above)
Comment (Polygonalr) [reply gcc/rust/lex/rust-lex.cc:1803]: There are subtle differences (different error messages, using `parse_escape` instead of `parse_utf8_escape`), so this was not abstracted out to a separate function.

File: gcc/rust/backend/rust-compile-type.cc [hunk 8]
```diff
@@ -684,6 +685,22 @@ TyTyResolveCompile::visit (const TyTy::ReferenceType &type)
 
       return;
     }
+  // Check for CStr, create a specific record for it
+  else if (type.is_dyn_cstr_type (&adt))
+    {
+      // &CStr is a fat pointer: { *const u8, usize }
```
Comment (CohenArthur) [inline gcc/rust/backend/rust-compile-type.cc]:
```rust
#[repr(transparent)]
CStr {
   inner: [u8]
}
```

**Review by CohenArthur — APPROVED** [review]
LGTM, sorry for the nitpick :)

**Review by CohenArthur — COMMENTED** [review]
Please add an execute test to ensure that the last character in the CStr is indeed a zero byte. Something like this:

```rust
fn main() -> i32 {
   let a = c"something";
   
   last_byte(a) as i32
}
```

**Review by philberty — APPROVED** [review]
LGTM
